### PR TITLE
ENH: replace `try/except ImportError` blocks with `if HAS_X`

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

Motivation: #16351 showed that we have hidden optional dependencies in a couple places, showing as modules that we *attempt* to import in a more-or-less recoverable fashion with plain try/except blocks such as
```python
try:
    import x
except ImportError:
    ...
else:
    ...
```
but that may not all be declared as optional dependencies in package metadata.

In this PR, I'm going for the bulk of these blocks (focusing on packages that are already declared as optional deps) and replacing them with our usual idiom
```python
if HAS_X:
    ...
else:
    ...
```
It some places, it leads to a user-visible improvement with how exceptions are chained, but the main idea is to make problematic occurrences of `except ImportError` like the one causing #16351 easier to spot.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### Approvals by sub-package

- [x] `convolution` https://github.com/astropy/astropy/pull/16353#pullrequestreview-2028644440
- [x] `coordinates` https://github.com/astropy/astropy/pull/16353#pullrequestreview-2028862899
- [x] `io.ascii` https://github.com/astropy/astropy/pull/16353#pullrequestreview-2028072991
- [ ] `io.fits`
- [x] `io.misc`
- [ ] `modeling`
- [ ] `nddata`
- [x] `stats` https://github.com/astropy/astropy/pull/16353#pullrequestreview-2028644440
- [x] `table` https://github.com/astropy/astropy/pull/16353#pullrequestreview-2028072991
- [x] `time` https://github.com/astropy/astropy/pull/16353#pullrequestreview-2028072991
- [ ] `timeseries`
- [ ] `utils`
- [x] `visualization` https://github.com/astropy/astropy/pull/16353#pullrequestreview-2028644440

### CODE REVIEW

File: astropy/io/fits/util.py
Comment (astrofrog) [reply astropy/io/fits/util.py:903]: Or does this PR avoid importing the optional deps to define HAS_?

Comment (astrofrog) [reply astropy/io/fits/util.py:903]: Ok then the comment above this can probably be removed/updated - thanks!

Comment (neutrinoceros) [reply astropy/io/fits/util.py:899]: done !

**Review by taldcroft — APPROVED** [review]
Good for io.ascii, table, time.

**Review by larrybradley — APPROVED** [review]
LGTM for `convolution`, `stats`, and `visualization`.

File: astropy/coordinates/solar_system.py
Comment (pllim) [inline astropy/coordinates/solar_system.py:169]: Can we maybe keep the error type the same?

Comment (eerovaher) [reply astropy/coordinates/solar_system.py:169]: [`ModuleNotFoundError`](https://docs.python.org/3/library/exceptions.html#ModuleNotFoundError) is a subclass of [`ImportError`](https://docs.python.org/3/library/exceptions.html#ImportError), so this change is backwards compatible.

**Review by eerovaher — APPROVED** [review]
I can approve for `coordinates`.

**Review by pllim — COMMENTED** [review]
In the places where `ImportError` is now `ModuleNotFoundError`, do they need a change log? Even if the new error is backward compatible, it is still a change, right?

Also in `stats`, looks like `ModuleNotFoundError` is newly introduced (or at least has a different message now). (But from the diff, not sure what it was before since it just assumed scipy was available.)

File: astropy/utils/xml/writer.py
Comment (mhvk) [inline astropy/utils/xml/writer.py:165]: Wonder if we should make this `ModuleNotFound` too...  Though fine not to in this PR.

**Review by mhvk — APPROVED** [review]
Thanks, that is a very nice cleanup. I feel the use of `ModuleNotFound` is an improvement too - and even suggest we use that in place of a `ValueError` in one case. 

But good to go as is!

File: astropy/utils/xml/writer.py
Comment (neutrinoceros) [reply astropy/utils/xml/writer.py:165]: Actually there are a couple other places where arbitrary exception types are used in place of `ImportError/ModuleNotFoundError`, but since changing types (for a non-subtype of the original) is a subtle API change, I'd prefer addressing them in a separate smaller PR, if that's alright with you.

**Comment by neutrinoceros** [comment]
> In the places where ImportError is now ModuleNotFoundError, do they need a change log? Even if the new error is backward compatible, it is still a change, right?

In my opinion it's borderline not user-visible (in the sense that I doubt anyone would notice) so I don't think it's worth a mention in the changelog, but obviously I'm happy to add one still if you guys think otherwise !

File: astropy/stats/funcs.py
Comment (pllim) [inline astropy/stats/funcs.py:1302]: This changes the message from:

`ImportError: cannot import name 'comb' from 'scipy.<module>' `

to:

`scipy is required for kuiper_false_positive_probability`

I'll defer to @larrybradley whether this needs a change log or not.

File: astropy/utils/xml/writer.py
Comment (mhvk) [reply astropy/utils/xml/writer.py:165]: Yes, fine to not do here.

File: astropy/stats/funcs.py
Comment (larrybradley) [reply astropy/stats/funcs.py:1302]: Technically, it is an API change, but I think it's an inconsequential one.  I doubt anyone is running this code and checking for the error type or message.

Comment (pllim) [reply astropy/stats/funcs.py:1302]: OK I think everyone else think we do not need a change log, so let's go with the majority. Thanks for the feedback!

**Comment by neutrinoceros** [comment]
Thanks @eerovaher for adding a bullet list to track approvals, and sorry I didn't anticipate that this PR would ping so many maintainers at once.

**Comment by pllim** [comment]
Re: pings -- It follows the rules at https://github.com/astropy/astropy/blob/main/.github/CODEOWNERS (based on opt-in by individual core maintainers). FYI.

**Comment by matteobachetti** [comment]
FWIW, changes in io.misc are fine

**Comment by nstarman** [comment]
Do we have enough approvals / has it been long enough waiting / can we merge this in?

**Comment by neutrinoceros** [comment]
Dunno, but better safe than sorry: rebased !

**Comment by neutrinoceros** [comment]
done

**Comment by pllim** [comment]
Thanks!

**Pull request merged**
